ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -216,6 +255,102 @@ KEXT_STATIC void UseMainForkIfNamedStream(
     }
 }
 
+static void ResizePendingRenames(uint32_t newMaxPendingRenames)
+{
+    PendingRenameOperation* newArray = Memory_AllocArray<PendingRenameOperation>(newMaxPendingRenames);
+    PendingRenameOperation* arrayToFree = nullptr;
+    uint32_t arrayToFreeLength = 0;
+    
+    SpinLock_Acquire(s_renameLock);
+    {
+        if (newMaxPendingRenames >= s_maxPendingRenames)
+        {
+            Array_CopyElements(newArray, s_pendingRenames, s_maxPendingRenames);
+            Array_DefaultInit(newArray + s_maxPendingRenames, newMaxPendingRenames - s_maxPendingRenames);
+            
+            arrayToFree = s_pendingRenames;
+            arrayToFreeLength = s_maxPendingRenames;
+            
+            s_pendingRenames = newArray;
+            s_maxPendingRenames = newMaxPendingRenames;
+        }
+    }
+    SpinLock_Release(s_renameLock);
+    
+    if (arrayToFree != nullptr)
+    {
+        assert(arrayToFreeLength > 0);
+        Memory_FreeArray(arrayToFree, arrayToFreeLength);
+    }
+}
+
+static void RecordPendingRenameOperation(vnode_t vnode)
+{
+    thread_t myThread = current_thread();
+    
+    bool resizeTable;
+    do
+    {
+        resizeTable = false;
+        uint32_t resizeTableLength = 0;
+        
+        SpinLock_Acquire(s_renameLock);
+        {
+            bool inserted = false;
+            for (uint32_t i = 0; i < s_maxPendingRenames; ++i)
+            {
+                if (s_pendingRenames[i].thread == nullptr)
+                {
+                    s_pendingRenames[i].thread = myThread;
+                    s_pendingRenames[i].vnode = vnode;
+                    inserted = true;
+                    break;
+                }
+                else
+                {
+                    assert(s_pendingRenames[i].thread != myThread);
```

Comment:
Assert cause Kernel panics right? 
Could this happen:
  1. Rename A on Thread 1
  2. Rename B on Thread 1
  3. PreDelete A on Thread 1
Want to verify how these events come in and are processed.

Reply:
This won't happen because a call with `KAUTH_FILEOP_WILL_RENAME` is always immediately followed by a `KAUTH_VNODE_DELETE`.